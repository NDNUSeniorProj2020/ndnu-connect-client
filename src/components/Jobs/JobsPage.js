import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import './Jobs.css';
import FilterJobsContainer from './FilterJobsContainer';
import { fetchJobs } from '../../actions/jobs/jobsActions';

export function ConnectedJobsPage({ jobs, success, fetchJobs }) {
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    const token = localStorage.getItem('token');

    fetchJobs(token);
    setIsLoading(false);
  }, [isLoading, fetchJobs]);

  if (success && !isLoading)
    return <FilterJobsContainer jobs={jobs} />;
  if (!success && !isLoading)
    return <p style={{ color: 'red' }}>Cannot load jobs.</p>;

  return <p>Loading page...</p>;
}

ConnectedJobsPage.propTypes = { jobs: PropTypes.array, success: PropTypes.bool, fetchJobs: PropTypes.func };
ConnectedJobsPage.defaultProps = { jobs: [], success: false, fetchJobs: f => f };

const mapStateToProps = ({ jobsReducer }) => ({ jobs: jobsReducer.jobs, success: jobsReducer.success });
const JobsPage = connect(mapStateToProps, { fetchJobs })(ConnectedJobsPage);

export default JobsPage;
