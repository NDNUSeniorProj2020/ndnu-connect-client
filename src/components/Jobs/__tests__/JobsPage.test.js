import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';

import JobsPage, { ConnectedJobsPage } from '../JobsPage';

// Define mock store
const mockStore = configureStore([]);

// Dummy data
const jobs = [
	{
		"id": 1,
		"title": "test1",
		"description": "description test",
		"qualifications": "bs degree",
		"pay": "1234",
		"link": "google.com",
		"date": "2020-03-26T11:59:26.227112-07:00",
		"type": "FULL",
		"user": 3
	}
];

describe('tests for JobsPage components', () => {
	describe('tests for ConnectedJobsPage', () => {
		describe('snapshot tests', () => {
			it('renders without crashing', () => {
				const tree = shallow(<ConnectedJobsPage />);
				expect(tree).toMatchSnapshot();
			});
		});

		describe('unit tests', () => {
			it('has defaultProps', () => {
				expect(ConnectedJobsPage.defaultProps.jobs).toEqual([]);
				expect(ConnectedJobsPage.defaultProps.fetchJobs('f')).toEqual('f');
			});
		});

		describe('integration tests', () => {
			let props;
			let useEffect;

			beforeEach(() => {
				useEffect = jest.spyOn(React, 'useEffect').mockImplementation(f => f());
				props = { fetchJobs: jest.fn().mockResolvedValue(jobs) };
			});

			afterEach(() => jest.clearAllMocks());

			it('calls fetchAuthors prop', () => {
				mount(
          <BrowserRouter>
            <ConnectedJobsPage {...props} success={true} jobs={jobs} />
          </BrowserRouter>
        );
				expect(props.fetchJobs).toHaveBeenCalled();
			});

			it('renders with error message if jobs cannt be fetched.', () => {
				const wrapper = mount(<ConnectedJobsPage {...props} />);
				expect(wrapper.find('p').text()).toEqual('Cannot load jobs.');
			});
		});
	});

	describe('tests for JobsPage component connected with Redux', () => {
		// Mock store setup
		let store;
		let component;

		beforeEach(() => {
			store = mockStore({
				jobsReducer: {
					jobs: [ ...jobs ],
					success: true
				}
			});
			store.dispatch = jest.fn();
			component = renderer.create(
				<Provider store={store}>
					<JobsPage />
				</Provider>
			);
		});

		describe('snapshot tests', () => {
			it('renders without crashing', () => expect(component).toMatchSnapshot());
		});
	});
});
