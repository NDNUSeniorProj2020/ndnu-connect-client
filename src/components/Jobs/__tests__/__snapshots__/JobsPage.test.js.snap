// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests for JobsPage components tests for ConnectedJobsPage snapshot tests renders without crashing 1`] = `ShallowWrapper {}`;

exports[`tests for JobsPage components tests for JobsPage component connected with Redux snapshot tests renders without crashing 1`] = `
<div>
  <div>
    <div>
      <div>
        <p>
          Job card
        </p>
      </div>
    </div>
  </div>
</div>
`;
